### ISSUE TITLE
Type of "date" could not be determined because it refers to itself when field has the same name as the type

### ISSUE THREAD

**Issue by ddorian** [microsoft/pyright#8213]
**Describe the bug**
We get an error when the field has the same name as the type. 

**Code or Screenshots**

```python
import dataclasses
from datetime import date


@dataclasses.dataclass()
class Test:
    date: date = dataclasses.field()

# my.py:7:11 - error: Type of "date" could not be determined because it refers to itself (reportGeneralTypeIssues)
# my.py:7:11 - error: Variable not allowed in type expression (reportInvalidTypeForm)
```

This code works fine though:
```python
import dataclasses
from datetime import date


@dataclasses.dataclass()
class Test:
    date2: date = dataclasses.field()

```

Using Python 3.12.4 &  pyright 1.1.368.

**Comment by erictraut** [comment]
Pyright's behavior here is correct. It reflects the runtime behavior. I therefore don't consider this a bug.

If you use an annotation within a class body that refers to a symbol defined in that class body, the python compiler interprets it as that symbol. You can see this as follows:

```python
class Foo:
    int: int = 3

print(Foo.__annotations__) # prints "{'int': 3}"
```

Here, the runtime uses the local (class-scoped) value of `int` rather than the global builtins value of `int` when evaluating the type annotation.

In any case, it's bad practice to use symbols in a way where the resolution is not obvious to a human reader, so I highly recommend against this practice even in cases where it happens to work both at runtime and in static type checkers. You'll find that other static type checkers and linters are less careful than pyright about honoring runtime behaviors, and even pyright may deviate from runtime behaviors in some undocumented corner cases.

The recommended practice is:
1. Avoid creating local symbols that shadow outer-scoped symbol names if possible.
2. If that's not possible, then create an alias for the outer-scoped symbol name and use it within the inner scope.

In your case, here's how I would write the code in my project. The meaning here is unambiguous to all human readers and will work with all static type checkers and linters.

```python
import dataclasses
from datetime import date as _date

@dataclasses.dataclass()
class Test:
    date: _date = dataclasses.field()
```
